File: rust/src/toml_codec.rs
//! The TOML codec: bytes to a structure, and a structure to canonical bytes.
//!
//! TOML is the one format wrench ships that cannot hold everything the other
//! two can, so two questions were answered rather than implemented around, and
//! both are answered the way the contract answers them elsewhere.
//!
//! A null is refused rather than substituted. TOML has no null, and FR-4.1 says
//! a value with no canonical form is refused rather than guessed at.
//!
//! A date decodes to its ISO 8601 string, which is what FR-2.9 requires of every
//! decoder. A TOML datetime carries which of the four temporal types it was, and
//! each is spelled the way Python's `tomllib` plus `.isoformat()` spells it, so
//! the packs agree about the same file.
//!
//! A TOML document is a table. There is no top-level scalar or array in the
//! format, so encoding one is refused rather than wrapped in an invented key.
//!
//! THE EMITTER IS WRITTEN BY HAND, for the reason the YAML emitter is: no TOML
//! library produces these bytes. Measured 2026-08-28, the same structure through
//! three libraries differed in array layout and in indentation under a table.

use serde_json::{Map, Value};

use crate::codec::Codec;

/// The TOML codec.
pub struct TomlCodec;

/// The one instance callers use.
pub const TOML: TomlCodec = TomlCodec;

type Fail = crate::Error;

impl Codec for TomlCodec {
    fn decode(&self, data: &[u8]) -> Result<Value, Fail> {
        let text = std::str::from_utf8(data).map_err(crate::Error::parse)?;
        let parsed: toml::Value = toml::from_str(text).map_err(crate::Error::parse)?;
        Ok(from_toml(&parsed))
    }

    fn encode(&self, value: &Value) -> Result<Vec<u8>, Fail> {
        let table = value
            .as_object()
            .ok_or_else(|| {
                crate::Error::encode("cannot write in canonical form: a TOML document is a table")
            })?;
        refuse_null(value, "")?;

        let mut out = String::new();
        write_table(&mut out, table, &[])?;
        Ok(out.into_bytes())
    }
}

/// A `toml::Value` as the maps, lists and JSON scalars FR-2.9 requires.
fn from_toml(value: &toml::Value) -> Value {
    match value {
        toml::Value::String(s) => Value::String(s.clone()),
        toml::Value::Integer(i) => Value::from(*i),
        toml::Value::Float(f) => Value::from(*f),
        toml::Value::Boolean(b) => Value::Bool(*b),
        // A datetime keeps which of the four types it was. `toml`'s Display
        // spells each one the way `tomllib` plus `.isoformat()` does, so the
        // packs agree without a format string per case.
        toml::Value::Datetime(d) => Value::String(d.to_string()),
        toml::Value::Array(items) => Value::Array(items.iter().map(from_toml).collect()),
        toml::Value::Table(table) => {
            let mut out = Map::new();
            for (name, item) in table {
                out.insert(name.clone(), from_toml(item));
            }
            Value::Object(out)
        }
    }
}

/// Refuse a null anywhere in the structure, naming where it sits.
fn refuse_null(value: &Value, where_: &str) -> Result<(), Fail> {
    match value {
        Value::Null => {
            let at = if where_.is_empty() {
                String::new()
            } else {
                format!(" at {where_}")
            };
            Err(crate::Error::encode(format!(
                "cannot write null in canonical form{at}: TOML has no null"
            )))
        }
        Value::Object(table) => {
            for (name, item) in table {
                let path = if where_.is_empty() {
                    name.clone()
                } else {
                    format!("{where_}.{name}")
                };
                refuse_null(item, &path)?;
            }
            Ok(())
        }
        Value::Array(items) => {
            for (index, item) in items.iter().enumerate() {
                refuse_null(item, &format!("{where_}[{index}]"))?;
            }
            Ok(())
        }
        _ => Ok(()),
    }
}

/// One table: its scalars, then its sub-tables, each sorted.
///
/// Scalars first because TOML binds a bare key to the most recent `[header]`, so
/// a scalar written after a sub-table would land inside it. That is a
/// correctness rule rather than a layout preference.
///
/// `serde_json`'s map is a `BTreeMap` here, because `preserve_order` is
/// deliberately not enabled, so iteration is already sorted.
fn write_table(out: &mut String, value: &Map<String, Value>, path: &[&str]) -> Result<(), Fail> {
    if !path.is_empty() {
        let parts: Vec<String> = path.iter().map(|part| key(part)).collect();
        out.push_str(&format!("[{}]\n", parts.join(".")));
    }

    let mut sections = Vec::new();
    for (name, item) in value {
        if item.is_object() || is_table_array(item) {
            sections.push(name);
            continue;
        }
        out.push_str(&format!("{} = {}\n", key(name), inline(item)?));
    }

    for name in sections {
        let mut deeper: Vec<&str> = path.to_vec();
        deeper.push(name);

        if let Some(table) = value[name].as_object() {
            out.push('\n');
            write_table(out, table, &deeper)?;
            continue;
        }
        // An array of tables is repeated `[[path]]` sections. TOML has no inline
        // form for one, so this is the only spelling available rather than a
        // choice between two.
        let parts: Vec<String> = deeper.iter().map(|part| key(part)).collect();
        let header = parts.join(".");
        for entry in value[name].as_array().expect("a table array is an array") {
            out.push_str(&format!("\n[[{header}]]\n"));
            write_table(out, entry.as_object().expect("a table array holds tables"), &[])?;
        }
    }
    Ok(())
}

/// A non-empty array whose every item is a table.
///
/// Empty stays inline as `[]`, because an empty array of tables and an empty
/// array of anything else are the same document and `[]` is the shorter of the
/// two spellings. A mixed array is not a table array and is refused by `inline`,
/// which is where the message about it belongs.
fn is_table_array(value: &Value) -> bool {
    match value.as_array() {
        Some(items) => !items.is_empty() && items.iter().all(Value::is_object),
        None => false,
    }
}

/// A key, bare where TOML allows it and quoted where it does not.
fn key(name: &str) -> String {
    if !name.is_empty()
        && name
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '-')
    {
        name.to_string()
    } else {
        string(name)
    }
}

/// A value on one line: a scalar, or an array of them.
fn inline(value: &Value) -> Result<String, Fail> {
    Ok(match value {
        Value::Bool(true) => "true".to_string(),
        Value::Bool(false) => "false".to_string(),
        Value::String(s) => string(s),
        Value::Number(n) => number(n),
        Value::Array(items) => {
            let parts: Result<Vec<String>, Fail> = items.iter().map(inline).collect();
            format!("[{}]", parts?.join(", "))
        }
        // An inline table would be a second way to spell a sub-table, and two
        // spellings of one thing is what canonical form exists to remove.
        Value::Object(_) => {
            return Err(crate::Error::encode(
                "cannot write a table inline in canonical form",
            ))
        }
        Value::Null => return Err(crate::Error::encode("cannot write null in canonical form")),
    })
}

/// A number, with a float in the spelling every codec shares. Kept in step with
/// the YAML and JSON codecs deliberately: a float spelled differently per format
/// is the same defect as one that reads back as an integer. FR-4.8.
fn number(n: &serde_json::Number) -> String {
    if let Some(i) = n.as_i64() {
        return i.to_string();
    }
    if let Some(f) = n.as_f64() {
        return crate::float_text::canonical_float_text(f);
    }
    n.to_string()
}

/// A basic string, escaped the way TOML spells escapes. FR-4.9.
///
/// TOML requires escaping the quote, the backslash and every control character
/// except tab, and offers far fewer names than YAML: no `\e`, no `\v`, no `\0`
/// and no `\x`. Anything without a name becomes `\uXXXX`, which is the format's
/// own spelling rather than a second answer to YAML's question.
///
/// C1 is deliberately left raw. TOML does not require escaping it and all three
/// parsers round trip it, so escaping would be this pack inventing a rule.
fn string(value: &str) -> String {
    let mut out = String::with_capacity(value.len() + 2);
    out.push('"');
    for ch in value.chars() {
        match ch {
            '\u{8}' => out.push_str("\\b"),
            '\t' => out.push_str("\\t"),
            '\n' => out.push_str("\\n"),
            '\u{c}' => out.push_str("\\f"),
            '\r' => out.push_str("\\r"),
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            c if c < ' ' || c == '\u{7f}' => out.push_str(&format!("\\u{:04X}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
    out
}

File: python/wrench/toml_codec.py
"""The TOML codec: bytes to a structure, and a structure to canonical bytes.

TOML is the one format wrench ships that cannot hold everything the other two
can, so two questions had to be answered rather than implemented around. Both
are answered the way the contract already answers them elsewhere.

**A null is refused rather than substituted.** TOML has no null. FR-4.1 says a
value with no canonical form is refused rather than guessed at, and that is the
same situation: writing `""` or omitting the key invents a document nobody
wrote, and a consumer reading it back gets a different structure than was saved.

**A date decodes to its ISO 8601 string.** FR-2.9 says a decoder produces maps,
lists and JSON scalars and nothing else. TOML has native date, time and datetime
types, exactly as YAML does, and the YAML codec already coerces those for the
same reason: the value reaches a JSON Schema validator that has no type for it,
and a structure that loads but cannot be saved back is what the rule prevents.
ISO 8601 is lossless for the value and is what the file already said.

**A TOML document is a table.** There is no top-level scalar or array in the
format, so encoding one is refused rather than wrapped in an invented key.
"""

from __future__ import annotations

import datetime
import tomllib

from wrench.codec import DELETE, FIRST_PRINTABLE, _normalise
from wrench.errors import EncodeError, ParseError, wrapping
from wrench.float_text import canonical_float_text


class TOMLCodec:
    """The format, knowing nothing about where the bytes came from."""

    def decode(self, data: bytes) -> object:
        """Bytes into maps, lists and scalars.

        `_normalise` is shared with the YAML codec rather than reimplemented,
        because the coercion rule is the contract's and not one format's: both
        formats have native temporal types and JSON has none.

        `tomllib.TOMLDecodeError` does not cross this boundary; the cause is
        kept and reachable.
        """
        with wrapping(ParseError):
            return _normalise(tomllib.loads(data.decode("utf-8")))

    def encode(self, value: object) -> bytes:
        """A structure into canonical bytes, keys sorted.

        WRITTEN BY HAND, for the reason the YAML emitter is: no TOML library
        produces these bytes. Measured 2026-08-28, the same structure through
        `tomli_w` and Go's `BurntSushi/toml` differed in array layout and in
        indentation under a table, so two packs using their own library disagree
        about canonical form, which is the failure wrench exists to prevent.

        The form: scalars first, sorted; then each sub-table as a `[path]`
        section, sorted; arrays inline on one line; no indentation inside a
        table.
        """
        if not isinstance(value, dict):
            raise EncodeError(None, f"cannot write {type(value).__name__} in canonical form: a TOML document is a table")
        with wrapping(EncodeError):
            _refuse_null(value, "")
            return "".join(_table(value, [])).encode("utf-8")


def _refuse_null(value: object, where: str) -> None:
    """Refuse a null anywhere in the structure, naming where it sits.

    Checked before encoding rather than left to the writer, because the writer's
    own message names a type rather than a path, and a caller with a nested
    document needs to find the key.
    """
    if value is None:
        at = f" at {where}" if where else ""
        raise ValueError(f"cannot write null in canonical form{at}: TOML has no null")
    if isinstance(value, dict):
        for key, item in value.items():
            _refuse_null(item, f"{where}.{key}" if where else str(key))
    elif isinstance(value, list):
        for index, item in enumerate(value):
            _refuse_null(item, f"{where}[{index}]")


def _table(value: dict, path: list[str]) -> list[str]:
    """One table: its scalars, then its sub-tables, each sorted.

    Scalars first because TOML binds a bare key to the most recent `[header]`,
    so a scalar written after a sub-table would land inside it. That is a
    correctness rule rather than a layout preference.
    """
    out: list[str] = []
    if path:
        out.append(f"[{'.'.join(_key(part) for part in path)}]\n")

    sections = []
    for name in sorted(value):
        item = value[name]
        if isinstance(item, dict) or _is_table_array(item):
            sections.append(name)
        else:
            out.append(f"{_key(name)} = {_inline(item)}\n")

    for name in sections:
        item = value[name]
        deeper = [*path, name]
        if isinstance(item, dict):
            out.append("\n")
            out.extend(_table(item, deeper))
            continue
        # An array of tables is repeated `[[path]]` sections. TOML has no inline
        # form for one, so this is the only spelling available rather than a
        # choice between two.
        header = ".".join(_key(part) for part in deeper)
        for entry in item:
            out.append(f"\n[[{header}]]\n")
            out.extend(_table(entry, [])[0:])
    return out


def _is_table_array(value: object) -> bool:
    """A non-empty list whose every item is a table.

    Empty stays inline as `[]`, because an empty array of tables and an empty
    array of anything else are the same document and `[]` is the shorter of the
    two spellings. A mixed list is not a table array and is refused by
    `_inline`, which is where the message about it belongs.
    """
    return bool(value) and isinstance(value, list) and all(isinstance(i, dict) for i in value)


def _key(name: object) -> str:
    """A key, bare where TOML allows it and quoted where it does not."""
    if not isinstance(name, str):
        raise TypeError(f"mapping key {name!r} is {type(name).__name__}, not a string")
    if name and all(c.isalnum() or c in "_-" for c in name):
        return name
    return _string(name)


def _inline(value: object) -> str:
    """A value on one line: a scalar, or an array of them."""
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return _number(value)
    if isinstance(value, str):
        return _string(value)
    if isinstance(value, (datetime.datetime, datetime.date, datetime.time)):
        return _string(value.isoformat())
    if isinstance(value, list):
        return "[" + ", ".join(_inline(item) for item in value) + "]"
    if isinstance(value, dict):
        # An inline table would be a second way to spell a sub-table, and two
        # spellings of one thing is what canonical form exists to remove.
        raise ValueError("cannot write a table inline in canonical form")
    raise ValueError(f"cannot write {type(value).__name__} in canonical form")


def _number(value: float) -> str:
    """A number, with a float in the spelling every codec shares.

    Kept in step with the YAML and JSON codecs deliberately: a float spelled
    differently per format is the same defect as one that reads back as an
    integer. FR-4.8.
    """
    if isinstance(value, int):
        return str(value)
    return canonical_float_text(value)


# TOML'S OWN ESCAPE TABLE, which is not YAML's. FR-4.9.
#
# TOML requires escaping the quote, the backslash and every control character
# except tab, and it offers far fewer names than YAML: there is no `\e`, no `\v`,
# no `\0` and no `\x`. So anything without a name here becomes `\uXXXX`, which is
# the format's own spelling rather than a second answer to YAML's question.
#
# C1 is deliberately left raw. TOML does not require escaping it and all three
# parsers round trip it, so escaping would be this pack inventing a rule.
_TOML_NAMED = {
    0x08: r"\b",
    0x09: r"\t",
    0x0A: r"\n",
    0x0C: r"\f",
    0x0D: r"\r",
    0x22: r"\"",
    0x5C: "\\\\",
}


def _string(value: str) -> str:
    r"""A basic string, escaped the way TOML spells escapes.

    Anything with a name gets it and every other control character becomes
    `\uXXXX`. C1 is left raw, which TOML permits and every parser round trips.
    """
    out = []
    for char in value:
        point = ord(char)
        named = _TOML_NAMED.get(point)
        if named is not None:
            out.append(named)
        elif point < FIRST_PRINTABLE or point == DELETE:
            out.append(f"\\u{point:04X}")
        else:
            out.append(char)
    return '"' + "".join(out) + '"'


TOML = TOMLCodec()
